django/core/exceptions.py

```diff
@@ -182,6 +184,30 @@ def __str__(self):
     def __repr__(self):
         return 'ValidationError(%s)' % self
 
+    def __eq__(self, other):
+        if not isinstance(other, ValidationError):
+            return NotImplemented
+        if hasattr(self, 'message'):
+            return (
+                hasattr(other, 'message') and
+                self.message == other.message and
+                self.code == other.code and
+                self.params == other.params
+            )   
+        return (
+            hasattr(self, 'error_list') == hasattr(other, 'error_list') and
+            hasattr(self, 'error_dict') == hasattr(other, 'error_dict') and
+            sorted(self.messages) == sorted(other.messages)
+        )
+
+    def __hash__(self):
+        if hasattr(self, 'message'):
+            return hash((make_hashable(self.message), make_hashable(self.code), make_hashable(self.params)))
+        if hasattr(self, 'error_list'):
+            return hash((make_hashable(sorted(self.messages)), 'error_list'))
```

Comment:
`mask_hashable()` returns a tuple or primitive so we don't need to wrap it into another tuple.

Reply:
I'm finding the number of brackets here difficult to follow. I'm trying to pass a tuple of `self.messages` and a string of 'error_list' into `hash()`. I'm adding the string to avoid an `error_list` returning the same hash as a `error_dict`

`hash()` only takes a single argument so need to wrap the two items in a tuple? 

Here is the same code with a bit more spacing to make the brackets easier to follow. I am not sure what I am missing. (sorry)

``` python
            return hash(
                (
                    make_hashable(sorted(self.messages)),
                    'error_dict'
                )
            )

```

Reply:
True, but I think we should take keys into account instead of adding `error_dict`/`error_list`.

Reply:
Sorry for dripping comments.

Reply:
No need for sorry, you area super hero 🦸!